File: barricade/src/main/java/com/mutualmobile/barricade/BarricadeInterceptor.java
package com.mutualmobile.barricade;

import java.io.IOException;
import java.util.Iterator;
import java.util.List;
import java.util.logging.Logger;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

/**
 * OkHttp interceptor that maps all outgoing requests to their barricaded responses (if available)
 */
public class BarricadeInterceptor implements Interceptor {

  private Barricade barricade = Barricade.getInstance();

  @Override public Response intercept(Chain chain) throws IOException {
    if (barricade.isEnabled()) {
      Request request = chain.request();
      String endpoint = request.url().encodedPath();
      String queryString = getParams(request);
      Response response = null;
      if(!queryString.isEmpty()){
        response = barricade.getResponseForParams(chain, endpoint,queryString);
      }else {
        response = barricade.getResponse(chain, endpoint);
      }
      if (response != null) {
        try {
          Thread.sleep(barricade.getDelay());
        } catch (InterruptedException e) {
          Logger.getLogger(Barricade.TAG).severe(e.getMessage());
        }
        return response;
      } else {
        Logger.getLogger(Barricade.TAG).severe("No response found, making actual request");
        return chain.proceed(request);
      }
    } else {
      return chain.proceed(chain.request());
    }
  }

  private String getParams(Request request){
    String requestType = request.method();
    switch (requestType.trim().toUpperCase()){
      case "PUT" :
      case "POST" : {
        //Todo : fetch parameters from post request
        return "";
      }
      case "GET" :
      default : {
       return getFormattedQuery(request);
      }
    }
  }


  private String getFormattedQuery(Request request){
    Iterator<String>  paramIterator = request.url().queryParameterNames().iterator();
    StringBuilder query = new StringBuilder();
    while(paramIterator.hasNext()){
      String paramName =paramIterator.next();
      List<String>values = request.url().queryParameterValues(paramName);
      for(String value: values){
        query.append(paramName).append("=").append(value).append("&");
      }
    }
    return query.toString();
  }
}

File: barricade/src/main/java/com/mutualmobile/barricade/BarricadeInterceptor.kt
package com.mutualmobile.barricade

import okhttp3.Interceptor
import okhttp3.Response
import java.util.logging.Logger

/**
 * OkHttp interceptor that maps all outgoing requests to their barricaded responses (if available)
 */
class BarricadeInterceptor : Interceptor {
    companion object {
        private val barricade = Barricade.getInstance()
    }

    override fun intercept(chain: Interceptor.Chain): Response {
        if (barricade.isEnabled) {
            val request = chain.request()
            val pathSegments = request.url.pathSegments
            val endpoint = pathSegments[pathSegments.lastIndex]
            val response = barricade.getResponse(chain, endpoint)
            response?.let {
                try {
                    Thread.sleep(barricade.delay)
                } catch (e: InterruptedException) {
                    Logger.getLogger(Barricade.TAG).severe(e.message)
                }
                return response
            } ?: run {
                Logger.getLogger(Barricade.TAG).severe("No response found, making actual request")
                return chain.proceed(request)
            }
        } else {
            return chain.proceed(chain.request())
        }
    }
}
